src/datomish/clauses.cljc

```diff
@@ -0,0 +1,256 @@
+      (util/concat-in
+        cc [:wheres]
+        (map
+          (fn [v]
+            (let [external (first (v theirs))
+                  internal (first (v ours))]
+              (assert external)
+              (assert internal)
+              [:= external internal]))
+          vars)))))
+
+(defn expand-where-from-bindings
+  "Take the bindings in the CC and contribute
+   additional where clauses. Calling this more than
+   once will result in duplicate clauses."
+  [cc]
+  (impose-external-bindings
+    (assoc cc :wheres (concat (bindings->where (:bindings cc))
+                              (:wheres cc)))))
+
+;; Pattern building is recursive, so we need forward declarations.
+(declare Not->NotJoinClause not-join->where-fragment)
+
+;; Accumulates a pattern into the CC. Returns a new CC.
+(defn apply-pattern-clause
+  "Transform a DataScript Pattern instance into the parts needed
+   to build a SQL expression.
+
+  @arg cc A CC instance.
+  @arg pattern The pattern instance.
+  @return an augmented CC"
+  [cc pattern]
+  (when-not (instance? Pattern pattern)
+    (raise-str "Expected to be called with a Pattern instance." pattern))
+  (when-not (instance? DefaultSrc (:source pattern))
+    (raise-str "Non-default sources are not supported in patterns. Pattern: " pattern))
+
+  (let [[table alias] (source->from (:source cc))   ; e.g., [:datoms :datoms123]
+        places (map (fn [place col] [place col])
+                    (:pattern pattern)
+                    (:columns (:source cc)))]
+    (reduce
+      (fn [cc
+           [pattern-part                           ; ?x, :foo/bar, 42
+            position]]                             ; :a
+        (let [col (sql/qualify alias (name position))]    ; :datoms123.a
+          (condp instance? pattern-part
+            ;; Placeholders don't contribute any bindings, nor do
+            ;; they constrain the query -- there's no need to produce
+            ;; IS NOT NULL, because we don't store nulls in our schema.
+            Placeholder
+            cc
+
+            Variable
+            (bind-column-to-var cc pattern-part col)
+
+            Constant
+            (constrain-column-to-constant cc col position (:value pattern-part))
+
+            (raise-str "Unknown pattern part " pattern-part))))
+
+      ;; Record the new table mapping.
+      (util/conj-in cc [:from] [table alias])
+
+      places)))
+
+(defn- plain-symbol->sql-predicate-symbol [fn]
+  (when-not (instance? PlainSymbol fn)
+    (raise-str "Predicate functions must be named by plain symbols." fn))
+  (#{:> :< :=} (keyword (name (:symbol fn)))))
+
+(defn apply-predicate-clause [cc predicate]
+  (when-not (instance? Predicate predicate)
+    (raise-str "Expected to be called with a Predicate instance." predicate))
+  (let [f (plain-symbol->sql-predicate-symbol (:fn predicate))]
+    (when-not f
+      (raise-str "Unknown function " (:fn predicate)))
```

Comment:
Just a reminder that DS has nice error raising code that includes the initial inputs; we should use that pattern.